plenum/server/future_primaries_batch_handler.py

```diff
@@ -1,30 +1,75 @@
-import copy
-from collections import OrderedDict
-from typing import Tuple, Dict
+from typing import Dict, List
 
 from common.exceptions import LogicError
-from plenum.common.constants import TXN_TYPE, NODE, TARGET_NYM, ALIAS, SERVICES, VALIDATOR, DATA, POOL_LEDGER_ID
-from plenum.common.util import getMaxFailures
+from plenum.common.constants import POOL_LEDGER_ID, \
+    AUDIT_LEDGER_ID, AUDIT_TXN_PRIMARIES, AUDIT_TXN_VIEW_NO
+from plenum.common.ledger import Ledger
+from plenum.common.txn_util import get_payload_data
 from plenum.server.batch_handlers.batch_request_handler import BatchRequestHandler
 from plenum.server.batch_handlers.three_pc_batch import ThreePcBatch
-from plenum.server.pool_manager import TxnPoolManager
 
 
 class FuturePrimariesBatchHandler(BatchRequestHandler):
     # This class is needed for correct primaries storing in audit ledger.
 
     def __init__(self, database_manager, node):
         super().__init__(database_manager, POOL_LEDGER_ID)
+        # map of view_no and list of primaries
+        self.primaries = {}    # type: Dict[int, List]
+        self.db_manager = database_manager
         self.node = node
 
+    def set_primaries(self, view_no, ps):
+        self.primaries[view_no] = ps
+
+    def _inspect_audit_txn(self, txn, view_no) -> int:
+        # Return delta of next primary's record
+        p_view_no = get_payload_data(txn)[AUDIT_TXN_VIEW_NO]
+        p_primaries = get_payload_data(txn)[AUDIT_TXN_PRIMARIES]
+        if p_view_no != view_no:
+            if isinstance(p_primaries, int):
+                return p_primaries
+            return 1
+        else:
+            if isinstance(p_primaries, int):
+                return p_primaries
+            return p_primaries
+
+    def _get_previous_primaries(self, audit, view_no, seq_no_end):
+        if seq_no_end == 0:
+            return None
+        previous_txn = audit.getBySeqNo(seq_no_end) if seq_no_end <= audit.size else audit.get_by_seq_no_uncommitted(seq_no_end)
```

Comment:
I believe you don't need a check `if seq_no_end <= audit.size`, you can call `get_by_seq_no_uncommitted` only, and will take care about everything.